### PR TITLE
SimDataFormats/ValidationFormats - move G4 dependency out of data format 

### PR DESCRIPTION
**Pull request by davidlange6**
Moved into TrackingMaterialProducer which already depends on G4.

### CODE REVIEW

File: SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc [hunk 1]
```diff
@@ -281,12 +283,24 @@ void TrackingMaterialProducer::update(const G4Step* step) {
   }
 
   // update track accounting
-  if (enter_sensitive)
-    m_track.enterDetector(sensitive, position, cosThetaPre);
+  if (enter_sensitive) {
+    if (m_track_volume != nullptr) {
+      edm::LogWarning("TrackingMaterialProducer") << "Entering volume " << sensitive << "while inside volume "
+                                                  << m_track_volume << ". Something is inconsistent";
```
Comment (vargasa) [inline SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc:289]: Hello @davidlange6. If this is inconsistent, is a warning enough?

File: SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc [hunk 2]
```diff
@@ -281,12 +283,24 @@ void TrackingMaterialProducer::update(const G4Step* step) {
...
+    m_track.enterDetector(position, cosThetaPre);
+  }
   m_track.step(MaterialAccountingStep(length, radiationLengths, energyLoss, globalPositionIn, globalPositionOut));
-  if (leave_sensitive)
-    m_track.leaveDetector(sensitive, cosThetaPost);
-
+  if (leave_sensitive) {
+    if (m_track_volume != sensitive) {
+      edm::LogWarning("TrackingMaterialProducer") << "Leaving volume " << sensitive << "while inside volume "
+                                                  << m_track_volume << ". Something is inconsistent";
```
Comment (vargasa) [inline SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc:299]: Should it be an error/exception?

**Review by vargasa — COMMENTED** [review]
Hello @davidlange6. Is it possible for you to update the same producer in the `dd4hep` folder please?

File: SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc (hunk 1 above)
Comment (davidlange6) [reply SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc:289]: No idea. it was a cerr which I wanted to make a message logger

File: SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc (hunk 2 above)
Comment (davidlange6) [reply SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc:299]: It wasn't before. No change in behavior is proposed...

**Comment by davidlange6** [comment]
> Hello @davidlange6. Is it possible for you to update the same producer in the `dd4hep` folder please?

can you be more specific? I don't see any needed change.

**Comment by vargasa** [comment]
> can you be more specific? I don't see any needed change.

Thank you @davidlange6 Please disregard my previous comment, this particular file didn't use any of the `DD/Core` classes and is not present in that folder

**Comment by davidlange6** [comment]
please test

File: SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc (hunk 2 above)
Comment (vargasa) [reply SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc:299]: I see, but this was handled internally by `SimDataFormats/ValidationFormats/src/MaterialAccountingTrack.cc`. I get a couple of [empty histograms](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-961ef0/8982/trackerMaterialAnalysisPlots/). FYI: @mmusich, @mtosi, @apsallid

**Comment by davidlange6** [comment]
It’s a condition that doesn’t happen. I’ve handled the error as it was done before unless ice misunderstood the code

Cheers,
David


On 28 Aug 2020, at 17:20, Andres Vargas <notifications@github.com> wrote:

﻿

@vargasa commented on this pull request.

________________________________

In SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc<https://github.com/cms-sw/cmssw/pull/31273#discussion_r479373188>:

>    m_track.step(MaterialAccountingStep(length, radiationLengths, energyLoss, globalPositionIn, globalPositionOut));
-  if (leave_sensitive)
-    m_track.leaveDetector(sensitive, cosThetaPost);
-
+  if (leave_sensitive) {
+    if (m_track_volume != sensitive) {
+      edm::LogWarning("TrackingMaterialProducer") << "Leaving volume " << sensitive << "while inside volume "
+                                                  << m_track_volume << ". Something is inconsistent";


I see, but this was handled internally by SimDataFormats/ValidationFormats/src/MaterialAccountingTrack.cc. I get a couple of empty histograms<https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-961ef0/8982/trackerMaterialAnalysisPlots/>. FYI: @mmusich<https://github.com/mmusich>, @mtosi<https://github.com/mtosi>, @apsallid<https://github.com/apsallid>

**Comment by civanch** [comment]
+1

**Comment by davidlange6** [comment]
@makortel - any feedback on this PR?

**Comment by smuzaffar** [comment]
+1
for BigProducts/Simulation

**Comment by qliphy** [comment]
+1

File: SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc (hunk 1 above)
Comment (vargasa) [reply SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc:289]: @davidlange6 by running:

`cmsRun SimTracker/TrackerMaterialAnalysis/test/trackingMaterialProducer10GeVNeutrino_ForPhaseI.py`

Now I get thousands of printouts like this:

```
Entering volume TECModule2RphiActivewhile inside volume TECModule1StereoActive. Something is inconsistent
Entering volume TECModule1RphiActivewhile inside volume TECModule2RphiActive. Something is inconsistent
Entering volume TECModule1StereoActivewhile inside volume TECModule1RphiActive. Something is inconsistent
Entering volume TECModule2RphiActivewhile inside volume TECModule1StereoActive. Something is inconsistent
Entering volume TECModule2RphiActivewhile inside volume TECModule2RphiActive. Something is inconsistent
Entering volume TECModule3RphiActivewhile inside volume TECModule2RphiActive. Something is inconsistent
Entering volume TECModule3RphiActivewhile inside volume TECModule3RphiActive. Something is inconsistent
Entering volume TECModule4StereoActivewhile inside volume TECModule3RphiActive. Something is inconsistent
Entering volume TECModule4RphiActivewhile inside volume TECModule4StereoActive. Something is inconsistent
Entering volume TECModule3RphiActivewhile inside volume TECModule4RphiActive. Something is inconsistent
Entering volume TECModule4StereoActivewhile inside volume TECModule3RphiActive. Something is inconsistent
Entering volume TECModule4RphiActivewhile inside volume TECModule4StereoActive. Something is inconsistent
Entering volume PixelBarrelActiveFull1while inside volume PixelBarrelActiveFull0. Something is inconsistent
Entering volume PixelBarrelActiveFull2while inside volume PixelBarrelActiveFull1. Something is inconsistent
Entering volume PixelBarrelActiveFull3while inside volume PixelBarrelActiveFull2. Something is inconsistent
Entering volume TIBActiveRphi0while inside volume PixelBarrelActiveFull3. Something is inconsistent
Entering volume TIBActiveSter0while inside volume TIBActiveRphi0. Something is inconsistent
Entering volume TIBActiveRphi0while inside volume TIBActiveSter0. Something is inconsistent
Entering volume TIBActiveSter0while inside volume TIBActiveRphi0. Something is inconsistent
Entering volume TIBActiveRphi2while inside volume TIBActiveSter0. Something is inconsistent
Entering volume TIBActiveRphi2while inside volume TIBActiveRphi2. Something is inconsistent
Entering volume TOBActiveRphi0while inside volume TIBActiveRphi2. Something is inconsistent
Entering volume TOBActiveRphi0while inside volume TOBActiveRphi0. Something is inconsistent
Entering volume TOBActiveSter0while inside volume TOBActiveRphi0. Something is inconsistent
Entering volume TOBActiveRphi0while inside volume TOBActiveSter0. Something is inconsistent
Entering volume TOBActiveRphi2while inside volume TOBActiveRphi0. Something is inconsistent
Entering volume TOBActiveRphi2while inside volume TOBActiveRphi2. Something is inconsistent
Entering volume PixelForwardSensorwhile inside volume PixelBarrelActiveFull0. Something is inconsistent
Entering volume PixelForwardSensorwhile inside volume PixelForwardSensor. Something is inconsistent
Entering volume PixelForwardSensorwhile inside volume PixelForwardSensor. Something is inconsistent
Entering volume PixelForwardSensorwhile inside volume PixelForwardSensor. Something is inconsistent
Entering volume PixelBarrelActiveFull1while inside volume PixelBarrelActiveFull0. Something is inconsistent
Entering volume PixelBarrelActiveFull2while inside volume PixelBarrelActiveFull1. Something is inconsistent
```
Can you please double check that these changes do not change the expected behavior? as I pointed out before merging histograms from the unit test are also empty. Thanks

Comment (davidlange6) [reply SimTracker/TrackerMaterialAnalysis/plugins/TrackingMaterialProducer.cc:289]: Fixed in #31752

**Pull request merged**